retry/retry.go

```diff
@@ -0,0 +1,97 @@
+// Package retry implements a generic and customizable retry mechanism.
+//
+// Took some inspiration from:
+// - https://github.com/eapache/go-resiliency/tree/main/retrier
+package retry
+
+import (
+	"errors"
+	"log/slog"
+	"time"
+)
+
+// Action is returned by a ClassifierFunc to indicate to Retry how to proceed.
+type Action int
+
+const (
+	// Success informs Retry that the attempt has been a success.
+	Success Action = iota
+	// HardFail informs Retry that the attempt has been a hard failure and
+	// thus should abort retrying.
+	HardFail
+	// SoftFail informs Retry that the attempt has been a soft failure and
+	// thus should keep retrying.
+	SoftFail
+)
+
+// Retry is the controller of the retry mechanism.
+// See the examples in file retry_example_test.go.
+type Retry struct {
+	UpTo         time.Duration // Total maximum duration of the retries.
+	FirstDelay   time.Duration // Duration of the first backoff.
+	BackoffLimit time.Duration // Upper bound duration of a backoff.
+	Log          *slog.Logger
+	SleepFn      func(d time.Duration) // Optional; used only to override in tests.
+}
+
+// BackoffFunc returns the next backoff duration; called by [Retry.Do].
+// You can use one of the ready-made functions [ConstantBackoff],
+// [ExponentialBackoff] or write your own.
+// Parameter err allows to inspect the error that caused the retry and return a
+// custom delay; this can be used in special cases such as when rate-limited
+// with a fixed window; see [github.Classifier] and [github.Backoff] for an
+// example.
+type BackoffFunc func(first bool, previous, limit time.Duration, err error) time.Duration
+
+// ClassifierFunc decides whether to proceed or not; called by [Retry.Do]
+type ClassifierFunc func(err error) Action
+
+// WorkFunc does the unit of work that might fail and need to be retried; called
+// by [Retry.Do].
+type WorkFunc func() error
+
+// Do is the loop of [Retry].
+// See the examples in file retry_example_test.go.
+func (rtr Retry) Do(
+	backoffFn BackoffFunc,
+	classifierFn ClassifierFunc,
+	workFn WorkFunc,
+) error {
+	if rtr.FirstDelay <= 0 {
+		return errors.New("FirstDelay must be positive")
+	}
+	if rtr.BackoffLimit <= 0 {
+		return errors.New("BackoffLimit must be positive")
+	}
+	if rtr.SleepFn == nil {
+		rtr.SleepFn = time.Sleep
+	}
+	rtr.Log = rtr.Log.With("system", "retry") // FIXME maybe better constructor???
```

Comment:
Can you explain the comment a bit? > FIXME maybe better constructor???
What needs to be fixed and why? in other words what you don't like here?

Reply:
@aliculPix4D 

There are two related aspects that I don't like:
- The current code is giving to the user a bare struct, `Retry`, for which the "zero values" of the fields are not usable.
- This more or less works because the only method of Retry is `Do`, which (exactly here in this piece of code) does the role of a constructor.

We have two options here:

1. Make the zero values usable. In this case, this would mean to use the zero values as an hint that the user wants the **default** values. So we would have to come up with reasonable default values (cannot be 0) for `FirstDelay` and `BackoffLimit`. On the other hand, the default value for `SleepFn` is clear: it is the real sleep, `time.Sleep`.
2. Introduce a `New` "constructor" that returns an error, instead of returning the error here in `Do`. In general, "fail fast".

Reply:
For details on the concept "make the zero value useful" (to which I would add: "when it is possible"), see https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful.

In general, we can argue that the lack of proper constructors in Go is more a defect than a merit, as this whole discussion indirectly shows...